Add speedometer to HUD

kindredtf opened the pull request:
Convars:
pf_speedo - enable/disable entire panel
pf_speedo_smoothing - enable/disable slight smoothing of values 
pf_speedo_bar - enable/disable graphical speed bar

Closes #239

kindredtf committed:
Add speedometer to HUD

Convars:
pf_speedo - enable/disable entire panel
pf_speedo_smoothing - enable/disable slight smoothing of values
pf_speedo_bar - enable/disable graphical speed bar

kindredtf committed:
Merge branch 'development' into addition/hud-speedo

kindredtf committed:
Add cpp to vpc script

kindredtf committed:
Merge branch 'addition/hud-speedo' of https://github.com/p4sstime/p4ss into hud-speedo

kindredtf committed:
Merge branch 'development' of https://github.com/p4sstime/p4ss into hud-speedo

drunderscore commented on game/p4ss/resource/ui/hudspeedo.res:
```diff
@@ -0,0 +1,108 @@
+		"teambg_2"		"../hud/misc_ammo_area_red"
+		"teambg_2_lodef"	"../hud/misc_ammo_area_red_lodef"
+		"teambg_3"		"../hud/misc_ammo_area_blue"
+		"teambg_3_lodef"	"../hud/misc_ammo_area_blue_lodef"				
```
Extra whitespace at the end of this line and elsewhere on other lines in the file.

drunderscore commented on src/game/client/tf/tf_hud_speedo.cpp:
```diff
@@ -0,0 +1,217 @@
+// $NoKeywords: $
+//=============================================================================//
+
+#include "cbase.h"
+#include "usermessages.h"
+#include "hud.h"
+#include "hudelement.h"
+#include "c_tf_player.h"
+#include "iclientmode.h"
+#include "ienginevgui.h"
+#include "tf_gamerules.h"
+#include "hud_macros.h"
+#include <vgui/ILocalize.h>
+#include <vgui/ISurface.h>
+#include <vgui/IVGui.h>
+#include <vgui_controls/EditablePanel.h>
+#include <vgui_controls/Label.h>
+#include "tf_shareddefs.h"
+#include <vgui_controls/ImagePanel.h>
```
Could keep it organized and alphabetize these <3

drunderscore commented on src/game/client/tf/tf_hud_speedo.cpp:
```diff
@@ -0,0 +1,217 @@
+using namespace vgui;
+
+ConVar pf_speedo( "pf_speedo", "1", FCVAR_ARCHIVE, "Enable/Disable the speedometer HUD element" );
+ConVar pf_speedo_smoothing( "pf_speedo_smoothing", "0", FCVAR_ARCHIVE, "Enable/disable smoothing for the HUD speedometer." );
```
I think smoothing looks _significantly_ better and should be enabled by default. Would be visually more appealing for most players and new players especially.

drunderscore reviewed (commented):
On the HUD, at or around max speed of `1500`, the fill bar will escape it's background bar. Probably should be visually clamped on the right with the same padding as the left?
<img width="196" height="100" alt="image" src="https://github.com/user-attachments/assets/3e3f57dc-6050-47a3-9f85-4545b90d4124" />

drunderscore committed:
Revert "Accounted for null terminator in MAX_PLAYER_NAME_LENGTH global variable - fixes Increase max player length to 32 chars (instead of 31) p4sstime/p4ss#121"

This reverts commit 6dbb22ca166e298c947fa9e50642e88adcbb48d1.

This inadvertently changed the layout of `player_info_s`, because one of
it's members references `MAX_PLAYER_NAME_LENGTH`. We can't do this,
because the engine references this structure, and those layouts cannot
differ!

Fixes #238 (likely twiddling with the `fakeplayer` member)

SolarLightTF2 committed:
New textures, borders, targetID, match hud, hudhealthaccount, many fixes

Not super thoroughly tested, I want to go to sleep

New borders and textures added. Affects some places indirectly, like the winpanel
New TargetID
New hudhealthaccount (+health when picking up HP or regenning HP)
New notifications (press F to inspect). Couldn't quite get the team textures to load without applying the rounded corner/outline shit so I just used a flat color texture for the team colors

Updated timer to not break with tf_use_match_hud 0
Updated waiting for players text, now matches setup
Updated pass/lockon notifications to be smaller and match the score display
Moved and reduced size of killfeed to prevent conflicts and added a background to improve readability

Animation overrides added. Edit scripts/hudanimations_pf.txt to override animations from hudanimations_tf.txt.
Disabled screenshot reminder popup animation during deathcam
Disabled match hud respawn animation (clearly not supposed to be there, plus it's buggy even in vanilla TF2. Most huds kill it)

Match hud updates:
Player panels tall value increased to make room for respawn count which now displays properly
No longer breaks in 12v12, but 6v7 looks odd, not much I can do without code editing
Moved round counter next to the timer

Fixed cutoff on scoreboard "server time left" text

kindredtf committed:
Removed whitespace

kindredtf committed:
Add padding, enable smoothing, alphabetize include

kindredtf committed:
Merge branch 'development' into addition/hud-speedo

drunderscore reviewed (changes requested):
This debug assertion sounds important:
`ASSERT: game/client/hud.cpp:441: Hud element 'CHudSpeedo' doesn't have an entry 'HudSpeedo' in scripts/HudLayout.res`

It should probably be in that file, and be positioned/laid out using that too.

CarlJosell commented:
#346 is a duplicate branch with all fixes for this speedo to work. Im closing this now.
